File: scripts/launch_wechat_hook.py
"""
启动微信 Hook 注入（参考 sim-bot-node launch-wechat-hook.mjs，不修改子项目代码）

用法: python scripts/launch_wechat_hook.py
依赖: WECHAT_HK_DIR 或项目根/hk 或 sim-bot-node/hk 下的 x64 inject.exe
"""
from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from utils.weixin_locate import locate_weixin_exe
from utils.weixin_process import (
    get_running_weixin_executable,
    get_running_weixin_pid,
    should_quit_weixin_before_inject,
    stop_running_weixin,
)
from utils.hook_inject_config import hook_inject_config_json


def resolve_hk_root() -> Path:
    if os.getenv('WECHAT_HK_DIR'):
        return Path(os.getenv('WECHAT_HK_DIR')).resolve()
    for candidate in (
        ROOT / 'hk',
        ROOT / 'sim-bot-node' / 'hk',
    ):
        if (candidate / 'x64 inject.exe').is_file():
            return candidate.resolve()
    return (ROOT / 'hk').resolve()


def main() -> int:
    if os.name != 'nt':
        print('本脚本仅支持 Windows。', file=sys.stderr)
        return 1

    hk_root = resolve_hk_root()
    inject_exe = hk_root / 'x64 inject.exe'
    dll_path = hk_root / 'libGLESv1.dll'

    if not inject_exe.is_file():
        print(f'未找到注入程序: {inject_exe}', file=sys.stderr)
        print('请设置 WECHAT_HK_DIR 或将 hk 套件放到项目 hk/ 目录', file=sys.stderr)
        return 1
    if not dll_path.is_file():
        print(f'未找到 DLL: {dll_path}', file=sys.stderr)
        return 1

    always_launch = os.getenv('WECHAT_INJECT_ALWAYS_LAUNCH') == '1'
    cached_exe: str | None = os.getenv('WECHAT_EXE_PATH', '').strip() or None
    if cached_exe and not Path(cached_exe).is_file():
        cached_exe = None

    if not always_launch and should_quit_weixin_before_inject() and get_running_weixin_pid():
        if not cached_exe:
            cached_exe = get_running_weixin_executable()
        print('[bootstrap] 关闭已运行的微信以便重新注入…')
        stop_running_weixin(wait_ms=3000)

    running_pid = None if always_launch else get_running_weixin_pid()
    target_arg: str

    if running_pid is None:
        print('查找微信…')
        weixin = cached_exe or locate_weixin_exe()
        if not weixin:
            print(
                '未找到 Weixin.exe / WeChat.exe。请确认已安装 PC 版微信；'
                '或在 .env 中设置 WECHAT_EXE_PATH=你的微信安装路径',
                file=sys.stderr,
            )
            return 1
        print(f'找到 — {weixin}')
        target_arg = str(Path(weixin).resolve())
    else:
        print(f'检测到微信已在运行（PID {running_pid}），直接注入…')
        target_arg = str(running_pid)

    json_arg = hook_inject_config_json()
    print('注入配置:', json_arg)

    cmd = [str(inject_exe), target_arg, str(dll_path), json_arg]
    print('执行:', ' '.join(cmd[:3]), '<config-json>')

    proc = subprocess.run(cmd, cwd=str(ROOT), capture_output=True, text=True)
    if proc.stdout:
        print(proc.stdout)
    if proc.stderr:
        print(proc.stderr, file=sys.stderr)
    if proc.returncode != 0:
        print(f'注入失败，退出码 {proc.returncode}', file=sys.stderr)
        return proc.returncode

    print('微信 Hook 注入完成。')
    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: sim-bot-desktop-pack-out/sim-bot-app/scripts/launch-wechat-hook.mjs
/**
 * 启动微信 Hook 注入：查找 Weixin.exe → 调用 hk\x64 inject.exe
 *
 * 若检测到微信已在运行，则向 inject 传入 PID（仅注入，不再拉起新微信）。
 * 若需始终传入 Weixin.exe 路径（可能再启实例），设置 WECHAT_INJECT_ALWAYS_LAUNCH=1。
 *
 * 用法（在项目根目录）: node scripts/launch-wechat-hook.mjs
 * 依赖环境变量见 .env（TCP_PORT、HOOK_CALLBACK_PORT、HOOK_DLL_HTTP_PORT 等）
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadRuntimeEnv } from '../src/util/load_runtime_env.js';
import { hookInjectConfigJson } from '../src/hook/inject_config.js';
import { spawn } from 'child_process';
import os from 'os';
import { locateWeixinExe } from './weixin-locate.mjs';
import {
  getRunningWeixinPid,
  shouldQuitWeixinBeforeInject,
  stopRunningWeixin,
} from './weixin-process.mjs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const projectRoot = process.env.SIM_BOT_ROOT
  ? path.resolve(process.env.SIM_BOT_ROOT)
  : path.resolve(__dirname, '..');
loadRuntimeEnv(projectRoot);

function main() {
  if (os.platform() !== 'win32') {
    console.error('本脚本仅支持 Windows。');
    process.exit(1);
  }

  const hkRoot = process.env.WECHAT_HK_DIR
    ? path.resolve(process.env.WECHAT_HK_DIR)
    : path.join(projectRoot, 'hk');
  const injectExe = path.join(hkRoot, 'x64 inject.exe');
  const dllPath = path.join(hkRoot, 'libGLESv1.dll');

  if (!fs.existsSync(injectExe)) {
    console.error('未找到注入程序：', injectExe);
    process.exit(1);
  }
  if (!fs.existsSync(dllPath)) {
    console.error('未找到 DLL：', dllPath);
    process.exit(1);
  }

  const alwaysLaunch = process.env.WECHAT_INJECT_ALWAYS_LAUNCH === '1';
  if (!alwaysLaunch && shouldQuitWeixinBeforeInject() && getRunningWeixinPid()) {
    console.log('[bootstrap] 关闭已运行的微信，以便按当前配置重新注入（仅此一次）…');
    stopRunningWeixin({ waitMs: 3000 });
  }

  let runningPid = alwaysLaunch ? null : getRunningWeixinPid();
  let weixinAbs = null;

  if (runningPid == null) {
    console.log('查找微信…');
    const weixin = locateWeixinExe({
      onFullScanHint: () => {
        console.log('   （注册表与常见路径未命中，正在进行全盘搜索，可能需要较长时间…）');
      },
    });

    if (!weixin) {
      console.error('未找到 Weixin.exe，请确认已安装 PC 版微信（Weixin）。');
      process.exit(1);
    }

    console.log(`找到——${weixin}`);
    weixinAbs = path.resolve(weixin);
  } else {
    console.log(`检测到微信已在运行（PID ${runningPid}），跳过定位安装路径，直接注入…`);
  }

  const jsonArg = hookInjectConfigJson();

  const targetArg = runningPid != null ? String(runningPid) : weixinAbs;

  if (runningPid == null) {
    if (alwaysLaunch) {
      console.log('WECHAT_INJECT_ALWAYS_LAUNCH=1，将向 inject 传入 Weixin.exe 路径…');
    } else {
      console.log('未检测到运行中的微信，将向 inject 传入 Weixin.exe 路径（由 inject 拉起）…');
    }
  }

  console.log('启动…');
  // inject.exe 参数须为绝对路径（含 DLL），避免依赖工作目录；首参为已有进程时为十进制 PID
  const child = spawn(injectExe, [targetArg, dllPath, jsonArg], {
    cwd: path.dirname(injectExe),
    detached: true,
    stdio: 'ignore',
    windowsHide: false,
  });
  child.unref();
  console.log('完成');
  console.log('');
  console.log('注入参数摘要：');
  if (runningPid != null) {
    console.log('  目标          PID', runningPid);
  } else {
    console.log('  Weixin.exe   ', weixinAbs);
  }
  console.log('  inject       ', injectExe);
  console.log('  DLL          ', dllPath);
  console.log('  JSON         ', jsonArg);
}

main();
